src/envoy/auth/jwt.cc

```diff
@@ -102,138 +126,139 @@ const uint8_t *CastToUChar(const std::string &str) {
-  std::string pkey_der = Base64::decode(pkey_pem);
-  return EvpPkeyFromRsa(
-      bssl::UniquePtr<RSA>(
-          RSA_public_key_from_bytes(CastToUChar(pkey_der), pkey_der.length()))
-          .get());
-}
+  // It returns "OK" or the failure reason.
+  Status GetStatus() { return status_; }
 
-bssl::UniquePtr<BIGNUM> BigNumFromBase64UrlString(const std::string &s) {
-  std::string s_decoded = Base64UrlDecode(s);
-  if (s_decoded == "") {
-    return nullptr;
+  bssl::UniquePtr<EVP_PKEY> EvpPkeyFromStr(const std::string &pkey_pem) {
+    std::string pkey_der = Base64::decode(pkey_pem);
+    if (pkey_der == "") {
+      UpdateStatus(Status::PUBKEY_PEM_BAD_FORMAT);
+      return nullptr;
+    }
+    return EvpPkeyFromRsa(
+        bssl::UniquePtr<RSA>(
+            RSA_public_key_from_bytes(CastToUChar(pkey_der), pkey_der.length()))
+            .get());
   }
-  return bssl::UniquePtr<BIGNUM>(
-      BN_bin2bn(CastToUChar(s_decoded), s_decoded.length(), NULL));
-};
 
-bssl::UniquePtr<RSA> RsaFromJwk(const std::string &n, const std::string &e) {
-  bssl::UniquePtr<RSA> rsa(RSA_new());
-  if (!rsa) {
-    // Couldn't create RSA key.
-    return nullptr;
+  bssl::UniquePtr<EVP_PKEY> EvpPkeyFromJwk(const std::string &n,
+                                           const std::string &e) {
+    return EvpPkeyFromRsa(RsaFromJwk(n, e).get());
   }
-  rsa->n = BigNumFromBase64UrlString(n).release();
-  rsa->e = BigNumFromBase64UrlString(e).release();
-  if (!rsa->n || !rsa->e) {
-    // RSA public key field is missing.
-    return nullptr;
-  }
-  return rsa;
-}
 
-bssl::UniquePtr<EVP_PKEY> EvpPkeyFromJwk(const std::string &n,
-                                         const std::string &e) {
-  return EvpPkeyFromRsa(RsaFromJwk(n, e).get());
-}
+ private:
+  // It holds failure reason.
+  Status status_;
 
-const EVP_MD *EvpMdFromAlg(const std::string &alg) {
-  // may use
-  // EVP_sha384() if alg == "RS384" and
-  // EVP_sha512() if alg == "RS512"
-  if (alg == "RS256") {
-    return EVP_sha256();
-  } else {
-    return nullptr;
+  void UpdateStatus(Status status) {
+    if (status_ == Status::OK) {
+      status_ = status;
+    }
   }
-}
-
-bool VerifySignature(EVP_PKEY *key, const std::string &alg,
-                     const uint8_t *signature, size_t signature_len,
-                     const uint8_t *signed_data, size_t signed_data_len) {
-  bssl::UniquePtr<EVP_MD_CTX> md_ctx(EVP_MD_CTX_create());
-  const EVP_MD *md = EvpMdFromAlg(alg);
 
-  if (!md) {
-    return false;
-  }
-  if (!md_ctx) {
-    return false;
-  }
-  if (EVP_DigestVerifyInit(md_ctx.get(), nullptr, md, nullptr, key) != 1) {
-    return false;
-  }
-  if (EVP_DigestVerifyUpdate(md_ctx.get(), signed_data, signed_data_len) != 1) {
-    return false;
-  }
-  if (EVP_DigestVerifyFinal(md_ctx.get(), signature, signature_len) != 1) {
-    return false;
+  bssl::UniquePtr<EVP_PKEY> EvpPkeyFromRsa(RSA *rsa) {
+    if (!rsa) {
+      UpdateStatus(Status::PUBKEY_RSA_OBJECT_NULL);
```

Comment:
Almost right. Now, there are 3 cases that it will happen:
- when `RSA_public_key_from_bytes()` returned `nullptr`, which is a parse error of PEM public key
- when `RSA_new()` returned `nullptr`, which I guess should be treated as crash
- when `BigNumFromBase64UrlString()` returned `nullptr`, which is a pares error of "n" or "e" field of a JWK

So, I guess this status should be renamed like `PUBKEY_PARSE_ERROR`.